Reduce usages of Prelude/Data.Text.head

jhrcek opened the pull request:
First batch of fixes to https://github.com/haskell/haskell-language-server/issues/2514

This is going to be harder than I thought :P
There are more `head`s to come, but opening this just to get initial feedback on the general approach.

Also (due to my ocd) I had to fix bunch of hlint warnings which lead to meaningful simplifications.

Vague questions: how are you guys dealing with running tests locally?
I'm getting bunch of `fd:28: hPutBuf: resource vanished (Broken pipe)` when running `cabal test` locally..

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2519"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

jhrcek commented on plugins/hls-tactics-plugin/src/Wingman/Judgements.hs:
```diff
@@ -34,6 +34,9 @@ buildHypothesis
       | Just ty <- t
+      -- TODO what's the intention behind this check?
+      -- can it be replace with something like `not . isSymOcc`?
+      -- https://hackage.haskell.org/package/ghc-9.2.1/docs/GHC-Plugins.html#v:isSymOcc
```
Probably question for @isovector 
My intention is to get rid of this usage of `head`

isovector replied:
This is checking that the bindings in scope were created in source Haskell. GHC makes lots of bindings whose name start with `$` or other odd things.

You could replace it with `maybe False isAlpha $ headMay $ occNameString occ`

jhrcek commented on ghcide/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -706,9 +706,10 @@ suggestModuleTypo Diagnostic{_range=_range,..}
-      findSuggestedModules = map (head . T.words) . drop 2 . T.lines
-      proposeModule mod = ("replace with " <> mod, TextEdit _range mod)
+    , "Perhaps you meant"     `T.isInfixOf` _message =
+      -- TODO findSuggestedModules never extracts module name - fix the parsing logic
```
This parsing logic is definitely broken. If I'm not wrong, this will always extract the word "Perhaps" ?!

pepeiborra replied:
Did you check the tests?

jhrcek replied:
This doesn't seem to be tested at all.
Also no action is generated for these kinds of errors, probably because the "Perhaps you meant" in on the 2nd line of the message, so it gets dropped by the drop 2 part of the original impl.
Let me add test and fix this in separate PR.

jneira commented on ghcide/src/Development/IDE/GHC/Compat/Units.hs:
```diff
@@ -324,7 +325,7 @@ moduleUnit =
 #endif
 
-filterInplaceUnits :: [UnitId] -> [PackageFlag] -> ([UnitId], [PackageFlag])
+filterInplaceUnits :: NE.NonEmpty UnitId -> [PackageFlag] -> ([UnitId], [PackageFlag])
```
I would use types unqualified, hope it is the pattern for other data types

michaelpj replied:
Ditto, not sure this one should change.

pepeiborra commented on ghcide/src/Development/IDE/Plugin/CodeAction/ExactPrint.hs:
```diff
@@ -321,7 +322,7 @@ extendImportViaParent df parent child (L l it@ImportDecl{..})
           isParentOperator = hasParen parent
       when hasSibling $
-        addTrailingCommaT (head pre)
+        addTrailingCommaT (headNote "Development.IDE.Plugin.CodeAction.ExactPrint.extendImportViaParent" pre)
```
Remove

michaelpj replied:
Suggest instead changing the `not null` guard on line 316 into a pattern guard that matches on `pre` and extracts the head there.

pepeiborra commented on ghcide/src/Development/IDE/Plugin/CodeAction/ExactPrint.hs:
```diff
@@ -396,9 +397,9 @@ extendHiding symbol (L l idecls) mlies df = do
     then when hasSibling $ do
       addTrailingCommaT x
-      addSimpleAnnT (head lies) (DP (0, 1)) []
+      addSimpleAnnT (headNote "Development.IDE.Plugin.CodeAction.ExactPrint.extendHiding1" lies) (DP (0, 1)) []
```
Remove

pepeiborra commented on ghcide/src/Development/IDE/Plugin/CodeAction/ExactPrint.hs:
```diff
@@ -396,9 +397,9 @@ extendHiding symbol (L l idecls) mlies df = do
+      addSimpleAnnT (headNote "Development.IDE.Plugin.CodeAction.ExactPrint.extendHiding1" lies) (DP (0, 1)) []
       unless (null $ tail lies) $
-        addTrailingCommaT (head lies) -- Why we need this?
+        addTrailingCommaT (headNote "Development.IDE.Plugin.CodeAction.ExactPrint.extendHiding2" lies) -- Why we need this?
```
Remove

pepeiborra commented on ghcide/src/Development/IDE/Plugin/Completions.hs:
```diff
@@ -210,7 +211,7 @@ extendImportHandler :: CommandFunction IdeState ExtendImport
   res <- liftIO $ runMaybeT $ extendImportHandler' ideState edit
   whenJust res $ \(nfp, wedit@WorkspaceEdit {_changes}) -> do
-    let (_, List (head -> TextEdit {_range})) = fromJust $ _changes >>= listToMaybe . toList
+    let (_, List (headNote "Development.IDE.Plugin.Completions.extendImportHandler" -> TextEdit {_range})) = fromJust $ _changes >>= listToMaybe . toList
```
Remove

michaelpj replied:
I think this should just do nothing if the match fails. It's already inside a `whenJust` for the response coming in.

pepeiborra commented on ghcide/src/Development/IDE/Spans/Documentation.hs:
```diff
@@ -63,7 +64,9 @@ lookupKind env mod =
 getDocumentationTryGhc :: HscEnv -> Module -> Name -> IO SpanDoc
-getDocumentationTryGhc env mod n = head <$> getDocumentationsTryGhc env mod [n]
+getDocumentationTryGhc env mod n =
+    headNote "Development.IDE.Spans.Documentation.getDocumentationTryGhc"
```
Remove

michaelpj replied:
This seems potentially kind of bad! I think we should do something here. Just return a `Maybe`?

pepeiborra replied:
I think this is similar to `uses` in that we are local equational reasoning suffices to prove the absence of errors. You could generalise `getDocumentationsTryGhc` to take  `Traversable` and use a `NonEmpty` list in `getDocumentationTryGhc`

jneira commented:
> I'm getting bunch of fd:28: hPutBuf: resource vanished (Broken pipe) when running cabal test locally..

I got them in windows long time ago, what is your setup?

pepeiborra commented on ghcide/test/exe/Main.hs:
```diff
@@ -4886,7 +4886,9 @@ projectCompletionTests =
         compls <- getCompletions doc (Position 1 13)
-        let item = head $ filter ((== "ALocalModule") . (^. Lens.label)) compls
+        item <- case find (\c -> c ^. Lens.label == "ALocalModule") compls of
+            Nothing -> liftIO . assertFailure $ "No completion with label ALocalModule found in : " <> show compls
```
Do we enable -fno-ignore-assert ? I don't remember

pepeiborra commented on plugins/hls-tactics-plugin/src/Wingman/CaseSplit.hs:
```diff
@@ -105,5 +106,5 @@ splitToDecl fixity name ams = do
 iterateSplit :: AgdaMatch -> [AgdaMatch]
 iterateSplit am =
   let iterated = iterate (agdaSplit =<<) $ pure am
-   in fmap wildify . head . drop 5 $ iterated
+   in fmap wildify . (\xs -> atNote "Wingman.CaseSplit.iterateSplit" xs 5) $ iterated
```
I would just use !!

michaelpj replied:
Right, this one is a bit funny because we know the list is infinite. Hmm.

pepeiborra reviewed (changes requested):
Most test suites need TASTY_NUM_THREADS=1 to override the default behaviour that runs tests concurrently. The ghcide testsuite sets this implicitly I believe, but others don't. 

If that's not enough, just copy the command line from CI

michaelpj commented on ghcide/bench/lib/Experiments.hs:
```diff
@@ -474,7 +475,7 @@ runBench runSess b = handleAny (\e -> print e >> return badRun)
         output $ "Setting up document contents took " <> showDuration d
         -- wait again, as the progress is restarted once while loading the cradle
         -- make an edit, to ensure this doesn't block
-        let DocumentPositions{..} = head docs
+        let DocumentPositions{..} = headNote "Experiments.runBench" docs
```
Yes, and I think if we're going to do this we should replace it with a meaningful error, not just the position where it occurred (although that's somewhat useful).

We also could be more lenient about partial functions in things like tests and benchmarks.

pepeiborra replied:
We don't want to have any errors escaping out in this way in ghcide. 

But this particular call site is in the benchmark suite which is allowed to error out, and a position is more helpful than any description in that case.

michaelpj commented on ghcide/src/Development/IDE/Core/Shake.hs:
```diff
@@ -853,18 +854,18 @@ defineNoDiagnostics op = defineEarlyCutoff $ RuleNoDiagnostics $ \k v -> (Nothin
 use :: IdeRule k v
     => k -> NormalizedFilePath -> Action (Maybe v)
-use key file = head <$> uses key [file]
+use key file = headNote "Development.IDE.Core.Shake.use" <$> uses key [file]
```
As I commented on the issue, I think you could solve all of these at a stroke by making the underlying `uses` functions work on `NonEmpty`, which I think would be fine.

pepeiborra replied:
Yes, I support the generalization to any `Traversable` to use `NonEmpty` here

michaelpj commented on ghcide/session-loader/Development/IDE/Session.hs:
```diff
@@ -758,7 +761,7 @@ getDependencyInfo fs = Map.fromList <$> mapM do_one fs
 removeInplacePackages
     :: UnitId     -- ^ fake uid to use for our internal component
-    -> [UnitId]
+    -> NonEmpty UnitId
```
Perhaps simpler to just convert to a normal list before calling this - I don't think _this_ function needs to require its argument to be non-empty!

michaelpj commented on ghcide/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -1553,7 +1551,7 @@ rangesForBinding' b (L l (IEThingWith _ thing _  inners labels))
 
 -- | 'matchRegex' combined with 'unifySpaces'
-matchRegexUnifySpaces :: T.Text -> T.Text -> Maybe [T.Text]
+matchRegexUnifySpaces :: T.Text -> T.Text -> Maybe (NonEmpty T.Text)
```
I get why this is the return type, but it's funny to see `Maybe (NonEmpty a)` instead of `[a]`!

googleson78 replied:
It *is* funny, but in my experience it's actually the more convenient/"right" choice, whenever you want to do something special for the empty list case.

michaelpj commented on ghcide/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -945,9 +949,11 @@ suggestImportDisambiguation df (Just txt) ps@(L _ HsModule {hsmodImports}) fileC
+                   )
+                 | local
+                 , not (null targetsWithRestImports)
+                 , let mode = HideOthers (uncurry (:) (head targetsWithRestImports))
```
still got a head here

michaelpj commented on plugins/hls-floskell-plugin/src/Ide/Plugin/Floskell.hs:
```diff
@@ -49,7 +50,9 @@ findConfigOrDefault file = do
-      let gibiansky = head (filter (\s -> styleName s == "gibiansky") styles)
+      let gibiansky = case find (\s -> styleName s == "gibiansky") styles of
+            Just gibStyle -> gibStyle
+            Nothing -> error "Ide.Plugin.Floskell.findConfigOrDefault: Style with name 'gibiansky' not found"
```
Can we just not set the `appStyle` if this fails?

michaelpj commented on plugins/hls-tactics-plugin/src/Wingman/Machinery.hs:
```diff
@@ -342,7 +342,7 @@ lookupNameInContext name = do
     :: TacticsM CType
 getDefiningType = do
-  calling_fun_name <- fst . head <$> asks ctxDefiningFuncs
+  calling_fun_name <- asks (fst . headNote "Wingman.Machinery.getDefiningType" . ctxDefiningFuncs)
```
Just below we do something with `failure`, can we also use that branch if we fail here?

michaelpj commented on plugins/hls-tactics-plugin/src/Wingman/Simplify.hs:
```diff
@@ -35,9 +36,9 @@ pattern Lambda pats body <-
-             -- amount of gas we give to auto
+  = -- Do three passes; this should be good enough for the limited
+    -- amount of gas we give to auto
+    (\xs -> atNote "Wingman.Simplify.simplify" xs 3)
```
Another annoying infinite list.

Anton-Latukha commented:
*Nota bene*: I addressed:
https://github.com/haskell/haskell-language-server/blob/bd0046b7089e32e309f7b29ee6ec1a808ca1db86/ghcide/src/Development/IDE/Spans/Documentation.hs#L64-L67
In https://github.com/haskell/haskell-language-server/pull/2349, - would soon merge it.

michaelpj reviewed (approved):
Generally looks good!

In future it would make this sort of thing easier to review if you don't also apply hlint suggestions. It's tempting, but reviewer bandwith is a bottleneck for us. You can always do a trivial "fix hlint suggestions PR" :)

Pepe, I'm interpreting your "Remove" comments as "go back to `head` here, because due to the implicit `HasCallStack` it is equally informative to `headNote`", is that right?

pepeiborra commented:
> Pepe, I'm interpreting your "Remove" comments as "go back to `head` here, because due to the implicit `HasCallStack` it is equally informative to `headNote`", is that right?

Correct

jhrcek commented:
Thanks for reviews, I'll address your comments tomorrow.

jhrcek commented:
I decided to close this PR as it contains too many unrelated changes.
I'll open multiple smaller, easier to review and hopefully less controversial PRs, addressing most of your comments and suggestions.
